src/parse.rs

```diff
@@ -1,34 +1,80 @@
 use crate::kind::PathKind;
 use crate::Resolver;
-use smol_str::SmolStr;
 
 #[derive(Clone, Debug)]
 pub struct Request {
-    pub target: SmolStr,
-    pub query: SmolStr,
-    pub fragment: SmolStr,
-    pub(crate) kind: PathKind,
-    pub(crate) is_directory: bool,
+    target: Box<str>,
+    query: Option<Box<str>>,
+    fragment: Option<Box<str>>,
+    kind: PathKind,
+    is_directory: bool,
+}
```

Comment:
Let's not expose any fields as public, since all we need is a `&str` at the consuming end. We can change these definitions without any breaking changes in the near future.